core/src/channel/channel_sf.rs

```diff
@@ -28,58 +28,73 @@ impl ChannelSoundfont {
         ChannelSoundfont {
             soundfonts: Vec::new(),
             matrix: VoiceSpawnerMatrix::new(),
-            curr_bank: None,
-            curr_preset: None,
+            curr_bank: 0,
+            curr_preset: 0,
         }
     }
 
     pub fn set_soundfonts(&mut self, soundfonts: Vec<Arc<dyn SoundfontBase>>) {
         self.soundfonts = soundfonts;
-        self.curr_bank = None;
-        self.curr_preset = None;
-        self.rebuild_matrix(0, 0);
+        self.rebuild_matrix(self.curr_bank, self.curr_preset, true);
     }
 
     pub fn change_program(&mut self, bank: u8, preset: u8) {
-        self.rebuild_matrix(bank, preset);
+        self.rebuild_matrix(bank, preset, false);
     }
 
-    fn rebuild_matrix(&mut self, bank: u8, preset: u8) {
-        if self.curr_bank == Some(bank) && self.curr_preset == Some(preset) {
+    fn rebuild_matrix(&mut self, bank: u8, preset: u8, force_rebuild: bool) {
+        if self.curr_bank == bank && self.curr_preset == preset && !force_rebuild {
```

Comment:
I added the `force_rebuild` for these two cases:
- Soundfont is not changed but a program change event with the same values as before is sent (to avoid rebuilding the matrix with the same spawners)
- A new soundfont is loaded but the preset and bank values are unchanged (so we need to load the same program from the new soundfont and not the piano)

I did not know about `Arc::ptr_eq` and it sounds like it can be used in this case instead of `force_rebuild`, but there will still be an if statement in that code. I will change it to this.

Reply:
Thinking again, adding the `ptr_eq` function will not change this code much because it will need to be checked in the `set_soundfonts` function where we have access to both the old and new soundfont. I can't check in the `rebuild_matrix` function because we only have access to the current loaded soundfonts. So basically I can either:
- leave it as it is with force rebuild
- add the check for rebuilding with the same program somewhere else (for example in the channel code, when receiving a bank change or PC event, to not request a rebuild at all if the values are not changed)
- Make it always rebuild without any checks (can become very slow on certain MIDIs that change instruments)

Reply:
Personally, I'd lean towards removing the check from `rebuild_matrix` and moving it to the 2 places that invoke it.

So like `if self.curr_bank == bank && self.curr_preset == preset` in change_program
and something like this for set_soundfonts:
```rs
fn are_vecs_equal<T>(
    old: &Vec<Arc<T>>,
    new: &Vec<Arc<T>>,
) -> bool {
    // First, check if the lengths are the same
    if old.len() != new.len() {
        return false;
    }

    // Then, check each pair of elements using Arc::ptr_eq
    for (old_item, new_item) in old.iter().zip(new.iter()) {
        if !Arc::ptr_eq(old_item, new_item) {
            return false;
        }
    }

    true
}
```

Reply:
> Personally, I'd lean towards removing the check from rebuild_matrix and moving it to the 2 places that invoke it.

Oh true this is a better approach.

> something like this for set_soundfonts

I hadn't thought of adding such a check there but it should be useful in some cases so I am going to add it.

Will make those changes in the next commit.